HULKs/hulk
# Aliveness service and pepsi subcommand

## Introduced Changes

This PR adds an aliveness service for querying multiple status information from NAOs in the network. The aliveness servive is built with yocto and will be integrated in the upcoming HULKs-OS version (see the aliveness branch of HULKs/meta-hulks).

Additionally, an `aliveness` subcommand is added to pepsi for querying the aliveness information with different verbosities (`summary`, `servicese`, `battery`, `ids`, `all`). Using `-j` or `--json` the aliveness information can also be returned as JSON for further processing in e.g. status bars.

The user can furthermore set a timeout via `-t` or `--timeout` (defaulting to 200ms) and specify NAO addresses (e.g. `22` or `10.1.24.22`) for querying the aliveness information only from specific NAOs.

## ToDo / Known Issues

The beacon for querying the aliveness information is send via multicast and the answers are received via unicast.
Since the answers are from a different IP address, most firewalls will block them.

Hence, the user has to allow the incoming messages in their firewall settings, e.g. for ufw by adding the following rule:

```
ufw allow proto udp from 10.1.24.0/24
```

## Ideas for Next Iterations (Not This PR)

Dynamic tab completions for pepsi subcommands such as `pepsi upload` and `pepsi poweroff` using the aliveness information are already WIP, but have been factored out to a different PR.

## How to Test

First, execute `cargo build` in `tools/aliveness` after sourcing the SDK.
Copy the target to a NAO, ssh into it and execute the target.
Now you can try any of the pepsi aliveness subcommands, e.g.

```
./pepsi aliveness summary
./pepsi aliveness services --json
./pepsi aliveness battery --timeout 500
./pepsi aliveness all 32
```

Make sure your firewall allows the incoming responses as described above if you don't see any responses.
